### PR TITLE
Refactor Room Shield / Profile shield

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Refactored the way we update RoomSummary, to avoid locking main thread with live data observable.

### CODE REVIEW

File: matrix-sdk-android-rx/src/main/java/im/vector/matrix/rx/RxRoom.kt [hunk 1]
```diff
@@ -30,103 +28,22 @@ import im.vector.matrix.android.api.session.room.send.UserDraft
...
-                        summaryObservable,
-                        roomEncryptionTrustLevelObservable,
-                        BiFunction { summary, level ->
-                            summary.getOrNull()?.copy(
-                                    roomEncryptionTrustLevel = level.getOrNull()
-                            ).toOptional()
-                        }
-                )
-                .doOnNext { Timber.v("RX: final room summary emitted for ${it.getOrNull()?.roomId}") }
+                .startWith(room.roomSummary().toOptional())
```
Comment (ganfra) [inline matrix-sdk-android-rx/src/main/java/im/vector/matrix/rx/RxRoom.kt]: Use StartWithCallable please

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/DefaultCrossSigningService.kt [hunk 2]
```diff
@@ -56,9 +54,9 @@ internal class DefaultCrossSigningService @Inject constructor(
         private val deviceListManager: DeviceListManager,
         private val uploadSigningKeysTask: UploadSigningKeysTask,
         private val uploadSignaturesTask: UploadSignaturesTask,
-        private val cryptoCoroutineScope: CoroutineScope,
-        private val coroutineDispatchers: MatrixCoroutineDispatchers,
-        private val taskExecutor: TaskExecutor) : CrossSigningService, DeviceListManager.UserDevicesUpdateListener {
+        private val computeTrustTask: ComputeTrustTask,
+        private val taskExecutor: TaskExecutor,
+        val eventBus: EventBus) : CrossSigningService, DeviceListManager.UserDevicesUpdateListener {
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/DefaultCrossSigningService.kt]: private?

File: matrix-sdk-android-rx/src/main/java/im/vector/matrix/rx/RxRoom.kt (hunk 1 above)
Comment (BillCarsonFr) [reply matrix-sdk-android-rx/src/main/java/im/vector/matrix/rx/RxRoom.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt [hunk 3]
```diff
@@ -64,5 +64,5 @@ interface CrossSigningService {
                          otherDeviceId: String,
                          locallyTrusted: Boolean?): DeviceTrustResult
 
-    fun getTrustLevelForUsers(userIds: List<String>): RoomEncryptionTrustLevel
+    suspend fun getTrustLevelForUsers(userIds: List<String>): RoomEncryptionTrustLevel
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt]: This is not used anymore. Remove? (Also for the moment we do not expose suspendable fun in the API)

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import im.vector.matrix.android.api.crypto.RoomEncryptionTrustLevel
+import im.vector.matrix.android.api.extensions.orFalse
+import im.vector.matrix.android.api.session.crypto.crosssigning.MXCrossSigningInfo
+import im.vector.matrix.android.internal.crypto.store.IMXCryptoStore
+import im.vector.matrix.android.internal.task.Task
+import javax.inject.Inject
+
+internal interface ComputeTrustTask : Task<ComputeTrustTask.Params, RoomEncryptionTrustLevel> {
+    data class Params(
+            val userList: List<String>
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: Rename to `userIds`

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+internal class DefaultComputeTrustTask @Inject constructor(
+        val cryptoStore: IMXCryptoStore
+) : ComputeTrustTask {
+
+    override suspend fun execute(params: ComputeTrustTask.Params): RoomEncryptionTrustLevel {
+        val userIds = params.userList
+        val allTrusted = userIds
+                .filter { getUserCrossSigningKeys(it)?.isTrusted() == true }
+
+        val allUsersAreVerified = userIds.size == allTrusted.size
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: Should be in the else branch of the next if

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/SessionToCryptoRoomMembersUpdate.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package im.vector.matrix.android.internal.crypto.crosssigning
+
+data class SessionToCryptoRoomMembersUpdate(
+        val encryptedRoomMembersUpdate: String,
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/SessionToCryptoRoomMembersUpdate.kt]: members?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ShieldTrustUpdater.kt [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import io.realm.Realm
+import io.realm.RealmConfiguration
+import kotlinx.coroutines.launch
+import org.greenrobot.eventbus.EventBus
+import org.greenrobot.eventbus.Subscribe
+import java.util.concurrent.atomic.AtomicReference
+import javax.inject.Inject
+
+internal class ShieldTrustUpdater @Inject constructor(
+        val eventBus: EventBus,
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ShieldTrustUpdater.kt]: private

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/store/db/RealmCryptoStore.kt [hunk 8]
```diff
@@ -403,14 +406,14 @@ internal class RealmCryptoStore(private val realmConfiguration: RealmConfigurati
...
-                            .`in`(UserEntityFields.USER_ID, userIds.toTypedArray())
+                            .`in`(UserEntityFields.USER_ID, userIds.distinct().toTypedArray())
                 },
                 { entity ->
                     entity.devices.map { CryptoMapper.mapToModel(it) }
                 }
         )
         return Transformations.map(liveData) {
-            it.firstOrNull() ?: emptyList()
+            it.flatten()
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/store/db/RealmCryptoStore.kt]: 👍

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt [hunk 9]
```diff
@@ -49,7 +50,8 @@ internal class RoomSummaryMapper @Inject constructor(
                         keysClaimed = result.claimedEd25519Key?.let { mapOf("ed25519" to it) },
                         forwardingCurve25519KeyChain = result.forwardingCurve25519KeyChain
                 )
-            } catch (e: MXCryptoError) {
+            } catch (e: Throwable) {
+                Timber.d(e)
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt]: why d and not e?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/DefaultCrossSigningService.kt (hunk 2 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/DefaultCrossSigningService.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt (hunk 3 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt]: As not used we can remove for now. (i thought I heard we will soon expose suspendable )

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt (hunk 5 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: 👍

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt (hunk 9 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt]: it's not really an error i think, it's very common, sometimes you don't have the keys (and there is no way you could have them)

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ShieldTrustUpdater.kt (hunk 7 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ShieldTrustUpdater.kt]: yes

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/SessionToCryptoRoomMembersUpdate.kt (hunk 6 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/SessionToCryptoRoomMembersUpdate.kt]: actually it's a roomId no?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt (hunk 3 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/crosssigning/CrossSigningService.kt]: ok

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt (hunk 9 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/mapper/RoomSummaryMapper.kt]: ok

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        return if (allTrusted.isEmpty()) {
+            RoomEncryptionTrustLevel.Default
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: We do not use the legacy method here?

Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: No. But we have iconsistency though, the room will be blackshield but the profile will be red (profile uses legacy) :/
Still moving https://docs.google.com/document/d/1RFUN0YRrOs8QZwiblaUbs4wWR0BjA0hBJ27rKJ57Zvk/edit?skip_itp2_check=true&pli=1#heading=h.jvldvebnsc7f

Traked here: https://github.com/vector-im/riotX-android/issues/963

Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/crosssigning/ComputeTrustTask.kt]: yes, thanks

**Pull request merged**